crates/telio-traversal/src/upgrade_sync.rs

```diff
@@ -195,23 +308,70 @@ impl State {
             .await
         {
             Ok(true) => {
-                // TODO: in LLT-4858 check if we have already sent our own upgrade request
-                // and if so compare keys to decide who should win
+                let decision = match self.upgrade_requests.get(&Direction::Sent(*public_key)) {
+                    Some(v) => {
+                        if Self::is_expired(self.expiration_period, v) {
+                            telio_proto::Decision::Accepted
+                        } else if Some(&self.our_public_key)
+                            == winning_key([&self.our_public_key, public_key])
+                        {
+                            // We reject this request, because we sent our own which is not
+                            // yet expired and our key is winning - the other side should accept
+                            // our request instead.
+                            telio_proto::Decision::RejectedDueToConcurrentUpgrade
+                        } else {
+                            telio_proto::Decision::Accepted
+                        }
+                    }
+                    None => telio_proto::Decision::Accepted,
```

Comment:
I feel like this should be in different PR tbh

Reply:
That's an integral part of the rfc 😕 Why do you think this shouldn't be a part of this PR?

Reply:
You can move this to `winning_key` function commit